### DISCUSSION TITLE
Per-font settings

### DISCUSSION THREAD

**Discussion by qwerasd205** [ghostty-org/ghostty#2973] [Feature Requests, Ideas]
This is a proposal to allow configuring font settings *per specified font* in the fallback chain. This includes:
- Size
- Metrics adjustments (baseline, etc.)
- Features
- Variations
- Any other font-modifying setting I may be forgetting about

### Motivation
One word: Control.

Currently, if using two different fonts, you're at the mercy of the individual fonts to determine their relative sizes and proportions. This is not ideal when using languages with mixed scripts or just when trying to customize things to use a different font for, e.g., italics. Getting fonts to look *as best as possible* together requires full control over each font, and currently our config does not allow for this.

### Also: Nerd Font Symbols
Very much related to this, I think we should separate the nerd font symbols from our embedded JB Mono, instead use a symbols only nerd font and allow adjusting its scale (perhaps with a dedicated config option), since it's a common complaint that nerd font icons aren't as large as desired.

If, in the per-font adjustments, the builtin fonts could be referred to by generic names (e.g. `builtin`, `builtin-symbols`, `builtin-emoji`) that would allow the user to adjust them to their liking. If we introduced an `offset` adjustment (we should!) this would also allow behavior like centering emojis by adding a horizontal offset.

A lot of requested behaviors can be achieved generically with this proposal, it offers a high level of control for users who want to make everything *just so* without entailing significant technical debt for us, which feels ideal.

### Config syntax
That's the sticking point, isn't it? I'm not really sure what the cleanest way to approach this would be. Maybe (unfortunately) just something akin to the CSS [`font`](https://developer.mozilla.org/en-US/docs/Web/CSS/font) shorthand property?

### (A separate but related idea:) Take lessons from the web
The web has [`font-size-adjust`](https://developer.mozilla.org/en-US/docs/Web/CSS/font-size-adjust) to allow better control and harmonization of fallback font glyphs with the main intended font. This is good, and some form of this in Ghostty would be great!

### Related
- #1244
- #1903
---

Sorry if this proposal is a little disjointed. If you have any questions, I can clarify and expand on anything I didn't make clear.

**Comment by qwerasd205** [comment]
To expand on the config syntax problem: I'm very iffy about the idea of inventing a lot of syntax for the config in general. I feel like the config should either be extremely simple or else use a pre-established syntax. In an ideal world, we would have units and obvious expressions, so you could do something like `font.builtin_symbols.size = 150%` and `font.builtin_symbols.baseline -= 2pt`, but if we're gonna have this sort of config with a lot of syntax then there are two things:
1. We've gotta draw the line somewhere when it comes to the config<->cli flags correspondence.
2. We should not do it ad hoc, we need to make a well defined syntax with consideration about current and future uses, or else find an existing one that has already solved the relevant problems and is suitable for use.

**Comment by qwerasd205** [comment]
And as for the undesirably small nerd font (and other) glyphs, another thing would be allowing a (perhaps configurable per font or codepoint range) rendering mode that expands glyphs to fill their available space. I've played with this in the past, and I think it's what Kitty does for nerd font glyphs.

**Comment by norbusan** [comment]
Great proposal! I see problems with Japanese fonts when dlig is activated, but can only deactivate it globally.

**Comment by teneburu** [comment]
I need this as well, Arabic fonts are so random...

**Comment by fharper** [comment]
I'm coming from iTerm2, and I had a font size setting for the standard font, and a different one for non-ASCII font, so an option to have two font-size configs is needed for me also.

**Comment by kimminss0** [comment]
I need this to adjust the size of CJK fonts bigger than the default.

**Comment by HuaDeity** [comment]
Another case where per-font feature targeting is needed: `font-feature = cv10` applied globally triggers unintended CJK glyph substitutions in the fallback font. Details in #11427.
